src/community/ardrive_contract_oracle.ts

```diff
@@ -0,0 +1,155 @@
+	 * Uses contract reader at index 0 first then descends down the list
+	 */
+	private contractReaders: ContractReader[] = [
+		new VertoContractReader(),
+		new SmartweaveContractReader(this.arweave)
+		// new RedstoneContractReader(this.arweave)
+	];
+	private currentContractReader = initialContractReader;
+	private readContractAttempts = initialContractAttempts;
+
+	private communityContract?: CommunityContractData;
+	private contractPromise?: Promise<CommunityContractData>;
+
+	/** Sets the current contract reader to the next reader in descending order  */
+	fallbackToNextContractReader(): void {
+		const nextContractReaderIndex = this.currentContractReader + 1;
+
+		this.readContractAttempts = initialContractAttempts;
+		this.currentContractReader = nextContractReaderIndex;
+
+		console.log('Falling back to next contract reader..');
+	}
+
+	/**
+	 * Reads a smart contract with the current contract reader
+	 *
+	 * @remarks Will fallback to other contract readers when one fails
+	 */
+	async readContract(txId: string): Promise<unknown> {
+		let contract: unknown;
+
+		try {
+			// Get contract with current contract reader's readContract implementation
+			contract = await this.contractReaders[this.currentContractReader].readContract(txId);
+			return contract;
+		} catch (error) {
+			console.error(`Contract could not fetched: ${error}`);
+
+			this.readContractAttempts++;
+
+			if (this.readContractAttempts >= maxReadContractAttempts) {
+				// Max attempts for contract reader has been reached
+				if (this.currentContractReader === this.contractReaders.length - 1) {
+					// Current contract reader is the last fallback,  throw an error
+					throw new Error(
+						`Max contract read attempts has been reached on the last fallback contract reader..`
+					);
+				}
+
+				// Else fallback to next reader
+				this.fallbackToNextContractReader();
+			} else {
+				console.log('Retrying with current contract reader..');
+			}
+
+			// Retry read contract if no error was thrown
+			return this.readContract(txId);
+		}
+	}
+
+	/**
+	 * @returns the ArDrive Community Smartweave Contract
+	 * @throws when the Community Contract cannot be fetched or is returned as falsy
+	 * @throws when the Community Contract is returned in an unexpected shape
+	 */
+	public async getCommunityContract(): Promise<CommunityContractData> {
+		// Contract data already cached, return contract data
+		if (this.communityContract) {
+			return this.communityContract;
+		}
+
+		// Contract promise still resolving, return promise with contract data
+		if (this.contractPromise) {
+			return this.contractPromise;
+		}
+
+		// Begin new contract read; cast result to known ArDrive Community Contract type
+		this.contractPromise = this.readContract(communityTxId) as Promise<CommunityContractData>;
+
+		this.communityContract = await this.contractPromise;
+
+		if (!this.communityContract.settings || !this.communityContract.balances) {
```

Comment:
Here's where I'd like for us to just be able to rely on type casting at runtime to throw the error for us rather than manual inspection of the data.